### PR TITLE
balance attacker selection

### PR DESCRIPTION
**Pull request by RiccardoBarb**
Fix issue #4

### CODE REVIEW

File: reign.py [hunk 1]
```diff
@@ -116,15 +121,27 @@ def battle(self):
...
-        empire_neighbours = self.obj.query(f'Empire == "{empire}"').iloc[0].empire_neighbours
+
+        empire_list = self.obj.Empire.values.tolist()
+        un_empire, empire_weight = np.unique(empire_list, return_counts=True)
+        probability_ratio = max(empire_weight)/empire_weight
+
+        if len(un_empire) < self.threshold and any(probability_ratio > self.low_b):
+            empire_weight[probability_ratio > self.low_b] = max(empire_weight)/self.low_b
+        else:
+            pass
```
Comment (jacogasp) [inline reign.py]: A cosa serve questo else/pass?

File: reign.py [hunk 2]
```diff
@@ -116,15 +121,27 @@ def battle(self):
         """
 
         # Choose ∑
-        empire = random.choice(self.obj.Empire.values.tolist())
-        empire_neighbours = self.obj.query(f'Empire == "{empire}"').iloc[0].empire_neighbours
+
+        empire_list = self.obj.Empire.values.tolist()
+        un_empire, empire_weight = np.unique(empire_list, return_counts=True)
```
Comment (jacogasp) [inline reign.py]: Should `un_empire` be `unique_empires` and `empire_weight` `empire_weights`?

File: reign.py [hunk 3]
```diff
@@ -116,15 +121,27 @@ def battle(self):
...
+        probability_ratio = max(empire_weight)/empire_weight
+
+        if len(un_empire) < self.threshold and any(probability_ratio > self.low_b):
+            empire_weight[probability_ratio > self.low_b] = max(empire_weight)/self.low_b
+        else:
+            pass
+
+        empire=random.choices(un_empire, empire_weight)
```
Comment (jacogasp) [inline reign.py]: Should `empire` be called `empires`?
Use spaces `empire = random.choices(un_empire, empire_weight)`

File: reign.py [hunk 4]
```diff
@@ -116,15 +121,27 @@ def battle(self):
...
 
         # Choose ∑
-        empire = random.choice(self.obj.Empire.values.tolist())
-        empire_neighbours = self.obj.query(f'Empire == "{empire}"').iloc[0].empire_neighbours
+
+        empire_list = self.obj.Empire.values.tolist()
+        un_empire, empire_weight = np.unique(empire_list, return_counts=True)
+        probability_ratio = max(empire_weight)/empire_weight
+
+        if len(un_empire) < self.threshold and any(probability_ratio > self.low_b):
```
Comment (jacogasp) [inline reign.py]: This line needs a comment.
Does `len(un_empire)` means the size of the empire?
Why do we use `any(probability_ratio > self.low_b)`?

File: reign.py [hunk 5]
```diff
@@ -30,6 +31,10 @@ def __init__(self, pandas_obj, should_display_map=False):
         self.__msg_cache_handler = None
 
         self.battle_round = 1
+
+        self.threshold = threshold
```
Comment (jacogasp) [inline reign.py]: Shall we use a better name? What threshold? Shall we call it `empire_size_low_threshold`?

File: reign.py [hunk 6]
```diff
@@ -30,6 +31,10 @@ def __init__(self, pandas_obj, should_display_map=False):
...
+        self.low_b = low_b
```
Comment (jacogasp) [inline reign.py]: Shall we call it `lowest_empire_weight` or `min_empire_size`?

File: reign.py (hunk 3 above)
Comment (jacogasp) [reply reign.py]: Otherwise, can we define `empire = random.choices(un_empire, empire_weight)[0]`

File: reign.py (hunk 1 above)
Comment (RiccardoBarb) [reply reign.py]: in realtà a nulla: se l'if ritorna un false, la situazione dovrebbe rimanere invariata

File: reign.py (hunk 2 above)
Comment (RiccardoBarb) [reply reign.py]: yes

File: reign.py (hunk 4 above)
Comment (RiccardoBarb) [reply reign.py]: `len(un_empire)` is the amount of empires in play. 
`any(probability_ratio > self.low_b)` returns True if at least 1 of the empire weight exceed the lower boundary. To be precise the lower boundary is set as being 1/3 of the highest weight.

File: reign.py (hunk 5 above)
Comment (RiccardoBarb) [reply reign.py]: yes, sounds good to me!

File: reign.py (hunk 1 above)
Comment (jacogasp) [reply reign.py]: Direi allora che e possiamo togliere!

**Pull request merged**